File: leetcode/design/379_Design_Phone_Directory/codes/linked_list.py
class ListNode:
    def __init__(self, val=0):
        self.val = val
        self.next = None

class PhoneDirectory:
    def __init__(self, maxNumbers: int):
        self.isSlotAvailable = [True] * maxNumbers
        self.head = ListNode(-1)
        cur = self.head
        for i in range(maxNumbers):
            cur.next = ListNode(i)
            cur = cur.next

    def get(self) -> int:
        if not self.head.next:
            return -1
        node = self.head.next
        self.head.next = node.next
        self.isSlotAvailable[node.val] = False
        return node.val

    def check(self, number: int) -> bool:
        return self.isSlotAvailable[number]

    def release(self, number: int) -> None:
        if self.isSlotAvailable[number]:
            return
        node = ListNode(number)
        node.next = self.head.next
        self.head.next = node
        self.isSlotAvailable[number] = True

# Your PhoneDirectory object will be instantiated and called as such:
# obj = PhoneDirectory(maxNumbers)
# param_1 = obj.get()
# param_2 = obj.check(number)
# obj.release(number)

File: leetcode/design/379_Design_Phone_Directory/codes/linked_list.cpp
#include <vector>
using namespace std;

struct Node {
    int val;
    Node* next;
    Node(int v) : val(v), next(nullptr) {}
};

class PhoneDirectory {
    vector<bool> isSlotAvailable;
    Node* head;

public:
    PhoneDirectory(int maxNumbers) {
        isSlotAvailable = vector<bool>(maxNumbers, true);
        head = new Node(-1);
        Node* cur = head;
        for (int i = 0; i < maxNumbers; ++i) {
            cur -> next = new Node(i);
            cur = cur -> next;
        }
    }

    int get() {
        if (!head -> next) return -1;
        Node* node = head -> next;
        head -> next = node -> next;
        isSlotAvailable[node -> val] = false;
        int val = node -> val;
        delete node;
        return val;
    }

    bool check(int number) {
        return isSlotAvailable[number];
    }

    void release(int number) {
        if (isSlotAvailable[number]) return;
        Node* node = new Node(number);
        node -> next = head -> next;
        head -> next = node;
        isSlotAvailable[number] = true;
    }
};

/**
 * Your PhoneDirectory object will be instantiated and called as such:
 * PhoneDirectory* obj = new PhoneDirectory(maxNumbers);
 * int param_1 = obj -> get();
 * bool param_2 = obj -> check(number);
 * obj -> release(number);
 */